lib/Conversion/CalyxToFSM/MaterializeFSM.cpp

```diff
@@ -0,0 +1,189 @@
+  void runOnOperation() override;
+
+  /// Assigns the 'fsm.output' operation of the provided 'state' to enabled the
+  /// set of provided groups. If 'topLevelDone' is set, also asserts the
+  /// top-level done signal.
+  void assignStateOutputOperands(OpBuilder &b, StateOp stateOp,
+                                 bool topLevelDone = false) {
+    SmallVector<Value> outputOperands;
+    auto &enabledGroups = stateEnables[stateOp];
+    for (StringAttr group : referencedGroups)
+      outputOperands.push_back(
+          getOrCreateConstant(b, APInt(1, enabledGroups.contains(group))));
+
+    assert(outputOperands.size() == machineOp.getNumArguments() - 1 &&
+           "Expected exactly one value for each uniquely referenced group in "
+           "this machine");
+    outputOperands.push_back(getOrCreateConstant(b, APInt(1, topLevelDone)));
+    stateOp.ensureOutput(b);
+    auto outputOp = stateOp.getOutputOp();
+    outputOp->setOperands(outputOperands);
+  }
+
+  /// Extends every `fsm.return` guard in the transitions of this state to also
+  /// include the provided set of 'doneGuards'. 'doneGuards' is passed by value
+  /// to allow the caller to provide additional done guards apart from group
+  /// enable-generated guards.
+  void assignStateTransitionGuard(OpBuilder &b, StateOp stateOp,
+                                  SmallVector<Value> doneGuards = {}) {
+    auto &enabledGroups = stateEnables[stateOp];
+    for (auto groupIt : llvm::enumerate(referencedGroups))
+      if (enabledGroups.contains(groupIt.value()))
+        doneGuards.push_back(machineOp.getArgument(groupIt.index()));
+
+    for (auto transition : stateOp.transitions().getOps<fsm::TransitionOp>()) {
+      transition.ensureGuard(b);
+      auto guardOp = transition.getGuardReturn();
+      llvm::SmallVector<Value> guards;
+      llvm::append_range(guards, doneGuards);
+      if (guardOp.getNumOperands() != 0)
+        guards.push_back(guardOp.operand());
+
+      if (guards.empty())
+        continue;
+
+      b.setInsertionPoint(guardOp);
+      Value guardConjunction;
+      if (guards.size() == 1)
+        guardConjunction = guards.front();
+      else
+        guardConjunction = b.create<comb::AndOp>(transition.getLoc(), guards);
+      guardOp.setOperand(guardConjunction);
+    }
+  }
+
+  Value getOrCreateConstant(OpBuilder &b, APInt value) {
+    auto it = constants.find(value);
+    if (it != constants.end())
+      return it->second;
+
+    OpBuilder::InsertionGuard g(b);
+    b.setInsertionPointToStart(&machineOp.getBody().front());
+    auto constantOp = b.create<hw::ConstantOp>(machineOp.getLoc(), value);
+    constants[value] = constantOp;
+    return constantOp;
+  }
+
+  /// Maintain a set of all groups referenced within this fsm.machine.
+  GroupSet referencedGroups;
+
+  /// Maintain a relation between states and the groups which they enable.
+  DenseMap<fsm::StateOp, GroupSet> stateEnables;
```

Comment:
I don't think you need a SetVector for the value of stateEnables. It seems like the order isn't used, and another kind of set-like container would probably be much cheaper.

Reply:
The use of a SetVector here was intentional in that `assignStateOutputOperands` places strong assumptions on the order of the value in `stateEnables` being identical to that present in the top-level I/O. The introduction of `SetVector` here was in the first place motivated by non-deterministic errors following from the above.

Reply:
I can see how the iteration order matters for `referencedGroups`, but I believe the values in the `stateEnables` map are only ever `insert`ed into or checked for `contains`. Am I missing something?